lib/SPIRV/SPIRVReader.cpp

```diff
@@ -2500,8 +2500,46 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
       return mapValue(BV, transSPIRVBuiltinFromInst(
                               static_cast<SPIRVInstruction *>(BV), BB));
 
-    if (isBinaryShiftLogicalBitwiseOpCode(OC) || isLogicalOpCode(OC))
+    if (isBinaryShiftLogicalBitwiseOpCode(OC) || isLogicalOpCode(OC)) {
+      // llvm.bswap intrinsic is not supported by SPIR-V and is lowered to
+      // set of logical shift, "or" and "and" instructions by regularization
+      // pass. We need to check if such set of instructions can be present in
+      // current SPIR-V module and replace it by llvm.bswap intrinsic call
+      // during reverse translation.
+      // See: llvm/lib/CodeGen/IntrinsicLowering.cpp:LowerBSWAP
+      std::string VarName = BV->getName();
+      if (VarName.find("bswap.") != std::string::npos) {
+        if ((VarName.find("bswap.2") != std::string::npos &&
+             static_cast<SPIRVTypeInt *>(BV->getType())->getBitWidth() == 16) ||
+            (VarName.find("bswap.4") != std::string::npos &&
+             static_cast<SPIRVTypeInt *>(BV->getType())->getBitWidth() == 32) ||
+            (VarName.find("bswap.8") != std::string::npos &&
+             static_cast<SPIRVTypeInt *>(BV->getType())->getBitWidth() == 64))
+          BSwapOperandMap.emplace(
+              VarName,
+              transValue(static_cast<SPIRVBinary *>(BV)->getOperand(0), F, BB));
+        if (VarName.find("bswap.i16") != std::string::npos) {
+          IRBuilder<> Builder(BB);
+          return mapValue(BV, Builder.CreateIntrinsic(
+                                  Intrinsic::bswap, transType(BV->getType()),
+                                  BSwapOperandMap["bswap.2"]));
```

Comment:
This is just a nit: we may use integral value as a key of the map, for example bitwidth of swapped type.

Reply:
We can't use bitwidth as the key for the map, because there are may be several bswap intrinsics with the same bitwidth. Вut, in this case, there is also an error in the current implementation. Thanks for pointing that out.

Reply:
Thanks, done